exercises/circular-buffer/circular_buffer_test.cpp

```diff
@@ -0,0 +1,187 @@
+}
+
+TEST_CASE("a_read_frees_up_capacity_for_another_write") 
+{
+    circular_buffer::circular_buffer<value_type> buffer(1);
+
+    REQUIRE_NOTHROW(buffer.write(1));
+
+    value_type expected = 1;
+    REQUIRE(expected == buffer.read());
+
+    REQUIRE_NOTHROW(buffer.write(2));
+
+    expected = 2;
+    REQUIRE(expected == buffer.read());
+}
+
+TEST_CASE("read_position_is_maintained_even_across_multiple_writes") 
+{
+    circular_buffer::circular_buffer<value_type> buffer(3);
+
+    REQUIRE_NOTHROW(buffer.write(1));
+    REQUIRE_NOTHROW(buffer.write(2));
+
+    value_type expected = 1;
+    REQUIRE(expected == buffer.read());
+
+    REQUIRE_NOTHROW(buffer.write(3));
+
+    expected = 2;
+    REQUIRE(expected == buffer.read());
+
+    expected = 3;
+    REQUIRE(expected == buffer.read());
+}
+
+TEST_CASE("items_cleared_out_of_buffer_cant_be_read") 
+{
+    circular_buffer::circular_buffer<value_type> buffer(1);
+
+    REQUIRE_NOTHROW(buffer.write(1));
+
+    buffer.clear();
+
+    REQUIRE_THROWS_AS(buffer.read(), std::domain_error);
+}
+
+TEST_CASE("clear_frees_up_capacity_for_another_write") 
+{
+    circular_buffer::circular_buffer<value_type> buffer(1);
+
+    REQUIRE_NOTHROW(buffer.write(1));
+
+    buffer.clear();
+
+    REQUIRE_NOTHROW(buffer.write(2));
+
+    value_type expected = 2;
+    REQUIRE(expected == buffer.read());
+}
+
+TEST_CASE("clear_does_nothing_on_empty_buffer") 
+{
+    circular_buffer::circular_buffer<value_type> buffer(1);
+
+    buffer.clear();
+
+    REQUIRE_NOTHROW(buffer.write(1));
+
+    value_type expected = 1;
+    REQUIRE(expected == buffer.read());
+}
+
+TEST_CASE("overwrite_acts_like_write_on_non_full_buffer") 
+{
+    circular_buffer::circular_buffer<value_type> buffer(2);
+
+    REQUIRE_NOTHROW(buffer.write(1));
+
+    buffer.overwrite(2);
+
+    value_type expected = 1;
+    REQUIRE(expected == buffer.read());
+
+    expected = 2;
+    REQUIRE(expected == buffer.read());
+}
+
+TEST_CASE("overwrite_replaces_the_oldest_item_on_full_buffer") 
+{
+    circular_buffer::circular_buffer<value_type> buffer(2);
+
+    REQUIRE_NOTHROW(buffer.write(1));
+    REQUIRE_NOTHROW(buffer.write(2));
+
+    buffer.overwrite(3);
+
+    value_type expected = 2;
+    REQUIRE(expected == buffer.read());
+
+    expected = 3;
+    REQUIRE(expected == buffer.read());
+}
+
+TEST_CASE("overwrite_replaces_the_oldest_item_remaining_in_buffer_following_a_read") 
+{
+    circular_buffer::circular_buffer<value_type> buffer(3);
+
+    REQUIRE_NOTHROW(buffer.write(1));
+    REQUIRE_NOTHROW(buffer.write(2));
+    REQUIRE_NOTHROW(buffer.write(3));
+
+    value_type expected = 1;
+    REQUIRE(expected == buffer.read());
+
+    REQUIRE_NOTHROW(buffer.write(4));
+
+    buffer.overwrite(5);
+
+    expected = 3;
+    REQUIRE(expected == buffer.read());
+
+    expected = 4;
+    REQUIRE(expected == buffer.read());
+
+    expected = 5;
+    REQUIRE(expected == buffer.read());
+}
+#endif  // !EXERCISM_RUN_ALL_TESTS
```

Comment:
Add a new test for using something other than an int. Easiest thing is to just use a `string` type and repeat the final test case.

Instead of using integers, just steal some of the words from here: https://github.com/exercism/cpp/blob/master/exercises/anagram/anagram_test.cpp